api/src/main/java/org/pmiops/workbench/opsgenie/EgressRemediationService.java

```diff
@@ -0,0 +1,193 @@
+    Collection<List<DbEgressEvent>> eventsByWorkspace =
+        events.stream()
+            .filter(e -> e.getWorkspace() != null)
+            .collect(Collectors.groupingBy(e -> e.getWorkspace().getWorkspaceId()))
+            .values();
+
+    for (List<DbEgressEvent> partition : eventsByWorkspace) {
+      List<Instant> sortedEventCreationTimes =
+          partition.stream()
+              .map(e -> e.getCreationTime().toInstant())
+              .sorted()
+              .collect(Collectors.toList());
+
+      // Note: partitions cannot be empty, per above workspace partitioning.
+      Instant lastGroupStart = sortedEventCreationTimes.get(0);
+      incidentCount++;
+
+      for (Instant eventCreationTime : sortedEventCreationTimes) {
+        // If we receive multiple events for the same user/workspace within a short duration, group
+        // these together for the purposes of incident history analysis.
+        if (eventCreationTime.isAfter(lastGroupStart.plus(EGRESS_INCIDENT_MERGE_WINDOW))) {
+          incidentCount++;
+          lastGroupStart = eventCreationTime;
+        }
+      }
+    }
+
+    return incidentCount;
+  }
+
+  private static Optional<Escalation> matchEscalation(
+      @Nullable EgressAlertRemediationPolicy policy, int incidentCount) {
+    if (policy == null) {
+      return Optional.empty();
+    }
+    List<Escalation> descendingEscalations =
+        policy.escalations.stream()
+            .sorted(Comparator.comparingInt((Escalation e) -> e.afterIncidentCount).reversed())
+            .collect(Collectors.toList());
+    for (Escalation e : descendingEscalations) {
+      // We validate against duplicate afterIncidentCount's upstream. If somehow we still have
+      // duplicates at this point, we pick the last matching escalation, as the above sort is
+      // stable.
+      if (e.afterIncidentCount <= incidentCount) {
+        return Optional.of(e);
+      }
+    }
+    return Optional.empty();
+  }
+
+  private void suspendUserCompute(DbUser user, Duration duration) {
+    Timestamp suspendUntil = Timestamp.from(clock.instant().plus(duration));
+    userService.updateUserWithRetries(
+        u -> {
+          // Prevent the user from interacting with compute in Workbench
+          u.setComputeSecuritySuspendedUntil(suspendUntil);
+          return u;
+        },
+        user,
+        Agent.asSystem());
+
+    int stopCount = leonardoNotebooksClient.stopAllUserRuntimesAsService(user.getUsername());
+    log.info(String.format("stopped %d runtimes for user", stopCount));
+  }
+
+  private void disableUser(DbUser user) {
+    userService.setDisabledStatus(user.getUserId(), true);
+
+    // also stop any running compute, killing any active egress processes the user may have
+    int stopCount = leonardoNotebooksClient.stopAllUserRuntimesAsService(user.getUsername());
+    log.info(String.format("stopped %d runtimes for user", stopCount));
```

Comment:
nit, we can make this log identical as the one below, like `stopped %d runtimes and disabled user`